src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -1348,10 +1388,25 @@ static const uint32_t *dds_stream_read_uni (dds_istream_t * __restrict is, char
   return ops;
 }
 
-static inline const uint32_t *dds_stream_read_adr (uint32_t insn, dds_istream_t * __restrict is, char * __restrict data, const uint32_t * __restrict ops)
+static inline const uint32_t *dds_stream_read_adr (uint32_t insn, dds_istream_t * __restrict is, char * __restrict data, const uint32_t * __restrict ops, bool is_mutable_member)
 {
   void *addr = data + ops[1];
-  if (op_type_external (insn) && DDS_OP_TYPE (insn) != DDS_OP_VAL_STR)
+  if (op_type_optional (insn))
+  {
+    /* mutable optional members are only included in PL-CDR2 if the
+       member is present, so present defaults to true */
+    bool present = is_mutable_member ? true : dds_is_get1 (is);
+    if (!present)
+    {
+      if (DDS_OP_TYPE (insn) != DDS_OP_VAL_STR)
+        *((char **) addr) = NULL;
```

Comment:
Good point, the tests didn't cover that, so in addition to fixing the bug I've also added a test case. That revealed another (related) bug: when re-using a buffer for reading multiple samples of a type containing an optional string, the pointer was not set to `NULL` in case the member was not present in a sample after reading a sample in which the member is present. This is also fixed in d577bb2